### PR TITLE
Added multiple (recording) icons and change them on recording

### PR DESCRIPTION
**Pull request by delvh**
Additionally updated and simplified IconUtil API

### CODE REVIEW

File: src/main/java/envoy/client/ui/IconUtil.java [hunk 1]
```diff
@@ -27,7 +31,16 @@ private IconUtil() {}
 	 * @return the icon
 	 * @since Envoy Client v0.1-beta
 	 */
-	public static Image load(String path) { return new Image(IconUtil.class.getResource(path).toExternalForm()); }
+	public static Image load(String path) {
+		Image image;
+		try {
+			image = new Image(IconUtil.class.getResource(path).toExternalForm());
+		} catch (final NullPointerException e) {
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/ui/IconUtil.java:38]: I don't think it is a good idea to catch the exception here. We are not able to handle it in this place, so it should be catched by the caller if appropriate. As this is a runtime exception that can't really be handled anyway.

File: src/main/java/envoy/client/ui/IconUtil.java [hunk 2]
```diff
@@ -38,9 +51,93 @@ private IconUtil() {}
 	 * @since Envoy Client v0.1-beta
 	 */
 	public static Image load(String path, int size) {
-		return new Image(IconUtil.class.getResource(path).toExternalForm(), size, size, true, true);
+		Image image;
+		try {
+			image = new Image(IconUtil.class.getResource(path).toExternalForm(), size, size, true, true);
+		} catch (final NullPointerException e) {
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/ui/IconUtil.java:57]: See above.

File: src/main/java/envoy/client/ui/IconUtil.java [hunk 3]
```diff
@@ -38,9 +51,93 @@ private IconUtil() {}
...
+		}
+		return image;
 	}
 
+	/**
+	 * Loads a {@code .png} icon from the sub-folder {@code /icons/} of the resource
+	 * folder and scales it to 16px.<br>
+	 * The suffix {@code .png} is automatically appended.
+	 *
+	 * @param name the image name without the .png - suffix
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/ui/IconUtil.java:69]:
```suggestion
	 * @param name the image name without the .png suffix
```

File: src/main/java/envoy/client/ui/controller/ChatScene.java [hunk 4]
```diff
@@ -110,7 +110,8 @@ private void initialize() {
 		messageList.setCellFactory(MessageListCellFactory::new);
 		userList.setCellFactory(ContactListCellFactory::new);
 
-		settingsButton.setGraphic(new ImageView(IconUtil.load("/icons/settings.png", 16)));
+		settingsButton.setGraphic(new ImageView(IconUtil.loadDefault("settings", 16)));
+		voiceButton.setGraphic(new ImageView(IconUtil.loadDefaultThemeSensitive("", 24)));
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/ui/controller/ChatScene.java:114]: Is this string literal supposed to be empty?

File: src/main/java/envoy/client/ui/IconUtil.java (hunk 1 above)
Comment (delvh) [reply src/main/java/envoy/client/ui/IconUtil.java:38]: Well, as you can see, I am printing the erroneous path in case no image could be loaded and then I'm rethrowing it, meaning the program still crashes as intended

File: src/main/java/envoy/client/ui/IconUtil.java (hunk 2 above)
Comment (delvh) [reply src/main/java/envoy/client/ui/IconUtil.java:57]: Cannot add anything else. See above.

File: src/main/java/envoy/client/ui/controller/ChatScene.java (hunk 4 above)
Comment (delvh) [reply src/main/java/envoy/client/ui/controller/ChatScene.java:114]: No, that is an artifact of testing the case that no image could be loaded

**Comment by delvh** [comment]
That is also a good idea. Should I implement this now (in this branch)?

**Comment by delvh** [comment]
Additionally, to be fair, two of the "new" icons are just copies of the old ones that did not conform with the convention used for the other ones, while the old ones were not deleted. Especially the black forward icon is still missing.

File: src/main/java/envoy/client/ui/IconUtil.java (hunk 1 above)
Comment (CyB3RC0nN0R) [reply src/main/java/envoy/client/ui/IconUtil.java:38]: This is not beneficial in my opinion. The exception will appear inside the console anyway. Most sources agree on the fact that log and throw is a rather bad practice (see [here](https://stackify.com/best-practices-exceptions-java/) (Section 8), [here](https://www.loggly.com/blog/logging-exceptions-in-java/) (middle of the page) or [here](https://stackoverflow.com/questions/6639963/why-is-log-and-throw-considered-an-anti-pattern)).

File: src/main/java/envoy/client/ui/IconUtil.java [hunk 5]
```diff
@@ -38,9 +51,93 @@ private IconUtil() {}
...
+	 *          {@code IconUtil.loadDefault("abc")}
+	 */
+	public static Image loadDefault(String name) { return load("/icons/" + name + ".png"); }
+
+	/**
+	 * Loads a {@code .png} icon from the sub-folder {@code /icons/} of the resource
+	 * folder and scales it to a given size.<br>
+	 * The suffix {@code .png} is automatically appended.
+	 *
+	 * @param name the image name without the .png - suffix
```
[the remark this answers is not in the document]
Comment (CyB3RC0nN0R) [reply src/main/java/envoy/client/ui/IconUtil.java:85]: I saw some of the other instances, however Eclipse's search & replace function is suited better to solve this problem than GitHub's suggestions :)

File: src/main/java/envoy/client/ui/IconUtil.java (hunk 1 above)
Comment (delvh) [reply src/main/java/envoy/client/ui/IconUtil.java:38]: So, if I understand your resources correctly, you want me to just throw `new NullPointerException("image path: " + path, e);`?

**Comment by CyB3RC0nN0R** [comment]
> Using theme specific folders is also a good idea. Should I implement this now (in this branch)? 

Yes, that'd be nice.

File: src/main/java/envoy/client/ui/IconUtil.java (hunk 1 above)
Comment (CyB3RC0nN0R) [reply src/main/java/envoy/client/ui/IconUtil.java:38]: That is one way to do it. I would just remove the catch block for simplicity's sake, however I'm equally comfortable with both solutions.

**Comment by delvh** [comment]
One question remains: What about the Exception? There are only two possible sources how a `NullPointerException` can happen here:
1. The user deletes the icons somehow
2. The programmer uses a wrong file location
In both cases, we want to be informed of the missing path. Afterall, it might not even be bad to just NOT throw the NullPointerExcception at all, since a user still wants his program to work, even if the graphics are missing. In the curent implementation, the application is immediately closed in case of such an error.

File: src/main/java/envoy/client/ui/IconUtil.java [hunk 6]
```diff
@@ -104,16 +104,16 @@ public static Image load(String path, int size) {
...
-	 *             the .png - suffix
+	 *             the .png suffix
 	 * @return the loaded image
 	 * @since Envoy Client v0.1-beta
 	 * @apiNote let's take two sample images {@code abc_black.png} and
 	 *          {@code abc_white.png} in the folder {@code /icons/} and load them.
 	 *          <br>
 	 *          To do that theme sensitve, we only have to call
-	 *          {@code IconUtil.loadDefaultThemeSensitive("abc_")}
+	 *          {@code IconUtil.loadIconThemeSensitive("abc_")}
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/ui/IconUtil.java]: This is out of date.

File: src/main/java/envoy/client/ui/IconUtil.java [hunk 7]
```diff
@@ -134,9 +134,9 @@ public static Image load(String path, int size) {
 	 *          size 16.
 	 *          <br>
 	 *          To do that theme sensitve, we only have to call
-	 *          {@code IconUtil.loadDefaultThemeSensitive("abc_", 16)}
+	 *          {@code IconUtil.loadIconThemeSensitive("abc_", 16)}
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/ui/IconUtil.java]: This (and the two lines above it) is out of date.

File: src/main/java/envoy/client/ui/IconUtil.java [hunk 8]
```diff
@@ -163,11 +163,13 @@ public static Image load(String path, int size) {
 	/**
 	 * This method should be called if the display of an icon depends upon the
 	 * currently active theme.<br>
-	 * In case of the dark theme, the suffix {@code "white"} will be appended, else
-	 * the suffix {@code "black"} will be appended.
+	 * In case of a default theme, the string returned will be
+	 * returned ({@code dark/} or {@code light/}), else it will be empty.
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/ui/IconUtil.java:167]:
```suggestion
	 * In case of a default theme, the string returned will be
	 * ({@code dark/} or {@code light/}), otherwise it will be empty.
```

**Comment by CyB3RC0nN0R** [comment]
> 
> 
> One question remains: What about the Exception? There are only two possible sources how a `NullPointerException` can happen here:
> 
>     1. The user deletes the icons somehow
> 
>     2. The programmer uses a wrong file location
>        In both cases, we want to be informed of the missing path. Afterall, it might not even be bad to just NOT throw the NullPointerExcception at all, since a user still wants his program to work, even if the graphics are missing. In the current implementation, the application is immediately closed in case of such an error.

I assess the two possible causes for the exception as follows:

1. This cannot happen unless the user opens the JAR file and manually removes the icons. In that case, I don't see a problem with the application crashing. As you might imagine, the same thing applies to the removal of class files from JAR, which cannot possibly be handled in a graceful way.

2. We are talking exclusively about static resource paths here. This means, that this mistake is very likely to be noticed during development. Because of this I see it as highly unlikely that the application doesn't crash during development but during production. Dropping the exception without printing the stack trace might cause the mistake to be unnoticed. Of course we can just print out the stack trace (aka log the exception) without rethrowing, however I see little value in doing this for the aforementioned reason. In a similar vein we don't handle the runtime exception that could arise from the `LoginCredentials` initialization (the name of the hashing algorithm is passed as a string). There are reasons for those things to fail, however they do not happen coincidentally, which is why they can be ignored given the right circumstances.

In conclusion, we can either:

- Log the exception without rethrowing: clutter the code and protect the user from bugs that can only happen during development or with extreme stupidity that we do nothing about in other cases (see `LoginCredentials` and `SceneContext#load`)

- Keep the method as is passing the runtime exception up to the caller without any handling
